simpleflow/execute.py

```diff
@@ -0,0 +1,193 @@
+
+    """
+    def wrap_callable(func):
+        @functools.wraps(func)
+        def execute(*args, **kwargs):
+            command = 'simpleflow.execute'  # name of a module.
+            try:
+                output = subprocess.check_output([
+                    interpreter, '-m', command,  # execute module a script.
+                    get_name(func), format_arguments_json(*args, **kwargs),
+                ],
+                    # Redirect stderr to stdout to get traceback on error.
+                    stderr=subprocess.STDOUT,
+                )
+            except subprocess.CalledProcessError as err:
+                print err.output
+                excline = err.output.rsplit('\n', 2)[1]
+                try:
+                    excname, excmsg = excline.split(':', 1)
+                except ValueError:
+                    excname = excline
+                    excmsg = ''
+
+                exception = make_callable(excname.strip())
+                raise exception(excmsg.strip())
+            return json.loads(output)
+
+        # Not automatically assigned in python < 3.2.
+        execute.__wrapped__ = func
+        return execute
+    return wrap_callable
+
+
+def program(path=None, argument_format=format_arguments):
+    def wrap_callable(func):
+        @functools.wraps(func)
+        def execute(*args, **kwargs):
+            check_arguments(argspec, args)
+            check_keyword_arguments(argspec, kwargs)
+
+            command = path or func.func_name
+            return subprocess.check_output(
+                [command] + argument_format(*args, **kwargs))
+
+        argspec = inspect.getargspec(func)
+        # Not automatically assigned in python < 3.2.
+        execute.__wrapped__ = func
+        return execute
+    return wrap_callable
+
+
+def make_callable(funcname):
+    if '.' not in funcname:
+        module_name = '__builtin__'
+        object_name = funcname
+    else:
+        module_name, object_name = funcname.rsplit('.', 1)
+
+    module = __import__(module_name, fromlist=['*'])
+    try:
+        callable = getattr(module, object_name)
+    except AttributeError:
+        raise AttributeError('module {} has no attribute {}'.format(
+            module.__name__,
+            object_name,
+        ))
+    return callable
+
+
+if __name__ == '__main__':
+    import sys
+
+    if len(sys.argv) < 2:
+        raise ValueError(
+            'cannot execute callable: missing callable name and arguments')
+
+    funcname = sys.argv[1]
+
+    if len(sys.argv) == 3:
+        raw_arguments = sys.argv[2]
+        try:
+            arguments = json.loads(raw_arguments)
+        except:
+            raise ValueError('cannot load arguments from {}'.format(
+                raw_arguments))
+
+    callable = make_callable(funcname).__wrapped__
+    args = arguments.get('args', ())
+    kwargs = arguments.get('kwargs', {})
+    result = callable(*args, **kwargs)
+    print(json.dumps(result))
```

Comment:
I may have missed something, but I'd expect `json.loads` here to *deserialize* what we would have returned above as a string through `json.dumps`.. or maybe I totally misunderstand the idea of json serializing here. Maybe it will get clearer when I try to run the tests tomorrow ;)